cmd/build/buildv2.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/okteto/okteto/pkg/okteto"
+	"github.com/okteto/okteto/pkg/registry"
+	"github.com/okteto/okteto/pkg/types"
+)
+
+// BuildV2 builds the images defined by a manifest
+func (bc *Command) BuildV2(ctx context.Context, manifest *model.Manifest, options *types.BuildOptions) error {
+	toBuildSvcs := getToBuildSvcs(manifest, options)
+	if err := validateOptions(manifest, toBuildSvcs, options); err != nil {
+		if errors.Is(err, oktetoErrors.ErrNoServicesToBuildDefined) {
+			oktetoLog.Infof("skipping BuildV2 due to not having any svc to build")
+			return nil
+		}
+		return err
+	}
+
+	buildManifest := manifest.Build
+
+	for _, svcToBuild := range toBuildSvcs {
+		if options.EnableStages {
+			oktetoLog.SetStage(fmt.Sprintf("Building service %s", svcToBuild))
+		}
+
+		buildSvcInfo := buildManifest[svcToBuild]
+		if !okteto.Context().IsOkteto && buildSvcInfo.Image == "" {
+			return fmt.Errorf("'build.%s.image' is required if your context is not managed by Okteto", svcToBuild)
+		}
+
+		imageTag, err := bc.buildService(ctx, manifest, svcToBuild, options)
+		if err != nil {
+			return err
+		}
+		oktetoLog.Success("Image for service '%s' pushed to registry: %s", svcToBuild, imageTag)
+		if err := bc.SetServicetEnvVars(svcToBuild, imageTag); err != nil {
+			return err
+		}
+	}
+	if options.EnableStages {
+		oktetoLog.SetStage("")
+	}
+	return manifest.ExpandEnvVars()
+}
+func (bc *Command) buildService(ctx context.Context, manifest *model.Manifest, svcName string, options *types.BuildOptions) (string, error) {
+	buildSvcInfo := manifest.Build[svcName]
+
+	switch {
+	case shouldBuildFromDockerfile(buildSvcInfo) && shouldAddVolumeMounts(buildSvcInfo):
+		image, err := bc.buildSvcFromDockerfile(ctx, manifest, svcName, options)
+		if err != nil {
+			return "", err
+		}
+		buildSvcInfo.Image = image
+		return bc.addVolumeMounts(ctx, manifest, svcName, options)
+	case shouldBuildFromDockerfile(buildSvcInfo):
+		return bc.buildSvcFromDockerfile(ctx, manifest, svcName, options)
+	case shouldAddVolumeMounts(buildSvcInfo):
+		if okteto.IsOkteto() {
+			return bc.addVolumeMounts(ctx, manifest, svcName, options)
+		}
+
+	default:
+		oktetoLog.Infof("could not build service %s, due to not having Dockerfile defined or volumes to include", svcName)
+	}
+	return "", nil
+}
+
+func (bc *Command) buildSvcFromDockerfile(ctx context.Context, manifest *model.Manifest, svcName string, options *types.BuildOptions) (string, error) {
+	oktetoLog.Information("Building image for service '%s'", svcName)
+	isStackManifest := manifest.Type == model.StackType
+	buildSvcInfo := getBuildInfoWithoutVolumeMounts(manifest.Build[svcName], isStackManifest)
+
+	buildOptions := build.OptsFromBuildInfo(manifest.Name, svcName, buildSvcInfo, &types.BuildOptions{})
+
+	// Check if the tag is already on global/dev registry and skip
+	if build.ShouldOptimizeBuild(buildOptions.Tag) && !buildOptions.BuildToGlobal {
+		oktetoLog.Debug("found OKTETO_GIT_COMMIT, optimizing the build flow")
```

Comment:
nit: From a client perspective sounds a bit weird this debug. As a client, I shouldn't care how the check to be optimized the build is done. Just `optimizing the build flow` would be a better log probably.

If in the future the check is done in other way, this could be missleading